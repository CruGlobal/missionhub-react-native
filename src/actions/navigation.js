import { NavigationActions } from 'react-navigation';

// Performance issues: https://github.com/react-community/react-navigation/issues/608#issuecomment-328635042

export function navigatePush(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.navigate({
      routeName: screen,
      params: props,
    }));
  };
}

export function navigateBack(times, backParams) {
  return (dispatch) => {
    if (times && times > 1) {
      dispatch(NavigationActions.pop({ n: times, immediate: true }));
    } else {
      // backParams can contain { key: string, immediate: bool }
      dispatch(NavigationActions.back(backParams));
    }
  };
}

export function navigateReset(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({ routeName: screen, params: props }),
      ],
    }));
  };
}

// The reset home and reset login are handled by the login/logout auth actions
