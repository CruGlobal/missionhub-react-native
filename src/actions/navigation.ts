/* eslint-disable @typescript-eslint/no-explicit-any, import/named */

import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import {
  NavigationActions,
  StackActions,
  NavigationNavigateAction,
} from 'react-navigation';

import { FEED_ITEM_DETAIL_SCREEN } from '../containers/Communities/Community/CommunityFeedTab/FeedItemDetailScreen/FeedItemDetailScreen';
import { MAIN_TABS, PEOPLE_TAB, COMMUNITIES_TAB } from '../constants';
import { COMMUNITY_TABS } from '../containers/Communities/Community/constants';

import { loadHome } from './auth/userData';

export function navigatePush(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.push({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export function navigateBack(times?: number) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    if (times && times > 1) {
      dispatch(StackActions.pop({ n: times, immediate: true }));
    } else {
      dispatch(NavigationActions.back());
    }
  };
}

export const navigateReset = (screen: string, props = {}) =>
  resetStack([
    NavigationActions.navigate({ routeName: screen, params: props }),
  ]);

export const navigateNestedReset = (
  screens: {
    routeName: string;
    tabName?: string;
    params?: { [key: string]: any };
  }[],
) =>
  resetStack(
    screens.map(({ routeName, params, tabName }) =>
      NavigationActions.navigate({
        routeName,
        ...(params ? { params } : {}),
        ...(tabName
          ? {
              action: NavigationActions.navigate({
                routeName: tabName,
              }),
            }
          : {}),
      }),
    ),
    screens.length - 1,
  );

const resetStack = (actions: NavigationNavigateAction[], index = 0) => (
  dispatch: ThunkDispatch<{}, {}, AnyAction>,
) =>
  dispatch(
    StackActions.reset({
      index,
      key: null, // Reset root stack navigator
      actions,
    }),
  );

// The reset home and reset login are handled by the login/logout auth actions

export function navigateReplace(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.replace({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export const navigateToMainTabs = (tabName = PEOPLE_TAB) => (
  dispatch: ThunkDispatch<never, {}, AnyAction>,
) => {
  dispatch(loadHome());
  dispatch(navigateNestedReset([{ routeName: MAIN_TABS, tabName }]));
};

export const navigateToFeedItemComments = (
  feedItemId: string,
  communityId: string,
) => (dispatch: ThunkDispatch<{}, null, AnyAction>) => {
  dispatch(
    navigateNestedReset([
      {
        routeName: MAIN_TABS,
        tabName: COMMUNITIES_TAB,
      },
      {
        routeName: COMMUNITY_TABS,
        params: { communityId },
      },
      {
        routeName: FEED_ITEM_DETAIL_SCREEN,
        params: { feedItemId },
      },
    ]),
  );
};
