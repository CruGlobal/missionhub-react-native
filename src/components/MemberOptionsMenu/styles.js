import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  container: {
    width: 32,
    height: 32,
  },
});
