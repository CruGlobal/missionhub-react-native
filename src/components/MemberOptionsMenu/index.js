import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Alert } from 'react-native';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';

import { transferOrgOwnership } from '../../actions/organizations';
import PopupMenu from '../PopupMenu';
import { makeAdmin, archiveOrgPermission } from '../../actions/person';
import { navigateBack } from '../../actions/navigation';
import { getMyCommunities } from '../../actions/organizations';

@translate('groupMemberOptions')
class MemberOptionsMenu extends Component {
  leaveCommunity = async () => {
    const { dispatch, person, personOrgPermission } = this.props;

    await dispatch(archiveOrgPermission(person.id, personOrgPermission.id));
    dispatch(getMyCommunities());
    dispatch(navigateBack());
  };

  makeAdmin = () => {
    const { dispatch, person, personOrgPermission } = this.props;
    dispatch(makeAdmin(person.id, personOrgPermission.id));
  };

  removeAdmin = () => {
    //TODO: remove admin
  };

  makeOwner = () => {
    const { dispatch, organization, person } = this.props;

    dispatch(transferOrgOwnership(organization.id, person.id));
  };

  removeFromCommunity = () => {
    const { dispatch, person, personOrgPermission } = this.props;
    dispatch(archiveOrgPermission(person.id, personOrgPermission.id));
  };

  canLeaveCommunity = () => {
    const { t, iAmOwner, organization } = this.props;

    if (iAmOwner) {
      const onPress = () => {
        Alert.alert(
          t('ownerLeaveCommunityErrorMessage', { orgName: organization.name }),
          null,
          { text: t('ok') },
        );
      };
      return [{ text: t(`leaveCommunity.optionTitle`), onPress }];
    }

    return this.createOption('leaveCommunity', this.leaveCommunity);
  };

  createOption = (optionName, optionMethod, hasDescription) => {
    const {
      t,
      person: { full_name: personName },
      organization: { name: communityName },
    } = this.props;

    const onPress = () => {
      Alert.alert(
        t(`${optionName}.modalTitle`, { personName, communityName }),
        hasDescription ? t(`${optionName}.modalDescription`) : null,
        [
          {
            text: t('cancel'),
            style: 'cancel',
          },
          {
            text: t(`${optionName}.confirmButtonText`),
            onPress: optionMethod,
          },
        ],
      );
    };

    return [{ text: t(`${optionName}.optionTitle`), onPress }];
  };

  render() {
    const { myId, person, iAmAdmin, iAmOwner, personIsAdmin } = this.props;

    const personIsMe = myId === person.id;

    const showLeaveCommunity = personIsMe;
    const showMakeAdmin = !personIsMe && iAmAdmin && !personIsAdmin;
    const showRemoveAdmin = !personIsMe && iAmAdmin && personIsAdmin;
    const showMakeOwner = !personIsMe && iAmOwner;
    const showRemoveMember = !personIsMe && iAmAdmin;

    const props = {
      actions: [
        ...(showLeaveCommunity ? this.canLeaveCommunity() : []),
        ...(showMakeAdmin
          ? this.createOption('makeAdmin', this.makeAdmin, true)
          : []),
        ...(showRemoveAdmin
          ? this.createOption('removeAdmin', this.removeAdmin)
          : []),
        ...(showMakeOwner
          ? this.createOption('makeOwner', this.makeOwner, true)
          : []),
        ...(showRemoveMember
          ? this.createOption('removeMember', this.removeFromCommunity)
          : []),
      ],
      iconProps: {},
    };
    return <PopupMenu {...props} />;
  }
}

MemberOptionsMenu.propTypes = {
  myId: PropTypes.string.isRequired,
  person: PropTypes.object.isRequired,
  organization: PropTypes.object.isRequired,
  personOrgPermission: PropTypes.object.isRequired,
  iAmAdmin: PropTypes.bool,
  iAmOwner: PropTypes.bool,
  personIsAdmin: PropTypes.bool,
};

export default connect()(MemberOptionsMenu);
