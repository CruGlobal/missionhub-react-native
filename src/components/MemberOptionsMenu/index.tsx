/* eslint complexity: 0 */

import React, { Component } from 'react';
import { connect } from 'react-redux-legacy';
import { Alert } from 'react-native';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';

import {
  transferOrgOwnership,
  removeOrganizationMember,
} from '../../actions/organizations';
import PopupMenu from '../PopupMenu';
import {
  makeAdmin,
  removeAsAdmin,
  archiveOrgPermission,
} from '../../actions/person';
import { navigateBack } from '../../actions/navigation';
import { CommunityMemberPerson } from '../CommunityMemberItem/__generated__/CommunityMemberPerson';

import styles from './styles';

export const API_TRY_IT_NOW_ADMIN_OWNER_ERROR_MESSAGE =
  'You must log in for admin or owner permissions';

// @ts-ignore
@withTranslation('groupMemberOptions')
class MemberOptionsMenu extends Component<{
  person: CommunityMemberPerson;
  organization: { id: string; name: string };
  personOrgPermission: { id: string; permission: string };
}> {
  async componentWillUnmount() {
    // @ts-ignore
    if (this.leaveCommunityOnUnmount) {
      // @ts-ignore
      const { dispatch, person, personOrgPermission } = this.props;

      await dispatch(archiveOrgPermission(person.id, personOrgPermission.id));
    }
  }

  leaveCommunity = () => {
    // @ts-ignore
    const { dispatch } = this.props;

    // @ts-ignore
    this.leaveCommunityOnUnmount = true;
    dispatch(navigateBack());
  };

  updatePermissionHandleTryItNowError = async (
    // @ts-ignore
    action,
    // @ts-ignore
    tryItNowErrorMessageFunction,
  ) => {
    // @ts-ignore
    const { dispatch, t } = this.props;

    try {
      await dispatch(action);
    } catch (error) {
      const errorDetail =
        error.apiError &&
        error.apiError.errors &&
        error.apiError.errors[0].detail;
      const errorMessage =
        errorDetail && tryItNowErrorMessageFunction(errorDetail);

      if (
        errorMessage &&
        errorMessage.includes(API_TRY_IT_NOW_ADMIN_OWNER_ERROR_MESSAGE)
      ) {
        return Alert.alert(t('tryItNowAdminOwnerErrorMessage'));
      }

      throw error;
    }
  };

  makeAdmin = () => {
    // @ts-ignore
    const { person, personOrgPermission } = this.props;

    return this.updatePermissionHandleTryItNowError(
      makeAdmin(person.id, personOrgPermission.id),
      // @ts-ignore
      errorDetail => errorDetail.permission_id && errorDetail.permission_id[0],
    );
  };

  removeAsAdmin = () => {
    // @ts-ignore
    const { dispatch, person, personOrgPermission } = this.props;

    dispatch(removeAsAdmin(person.id, personOrgPermission.id));
  };

  makeOwner = () => {
    // @ts-ignore
    const { organization, person } = this.props;

    return this.updatePermissionHandleTryItNowError(
      transferOrgOwnership(organization.id, person.id),
      // @ts-ignore
      errorDetail => errorDetail,
    );
  };

  removeFromCommunity = async () => {
    // @ts-ignore
    const { dispatch, person, personOrgPermission, organization } = this.props;

    await dispatch(archiveOrgPermission(person.id, personOrgPermission.id));
    dispatch(removeOrganizationMember(person.id, organization.id));
  };

  canLeaveCommunity = () => {
    // @ts-ignore
    const { t, iAmOwner, organization } = this.props;

    if (iAmOwner) {
      const onPress = () => {
        Alert.alert(
          t('ownerLeaveCommunityErrorMessage', { orgName: organization.name }),
          // @ts-ignore
          null,
          { text: t('ok') },
        );
      };
      return [{ text: t('leaveCommunity.optionTitle'), onPress }];
    }

    // @ts-ignore
    return this.createOption('leaveCommunity', this.leaveCommunity);
  };

  // @ts-ignore
  createOption = (optionName, optionMethod, hasDescription) => {
    const {
      // @ts-ignore
      t,
      // @ts-ignore
      person: { fullName: personName },
      // @ts-ignore
      organization: { name: communityName },
    } = this.props;

    const onPress = () => {
      Alert.alert(
        t(`${optionName}.modalTitle`, { personName, communityName }),
        hasDescription ? t(`${optionName}.modalDescription`) : null,
        [
          {
            text: t('cancel'),
            style: 'cancel',
          },
          {
            text: t(`${optionName}.confirmButtonText`),
            onPress: optionMethod,
          },
        ],
      );
    };

    return [{ text: t(`${optionName}.optionTitle`), onPress }];
  };

  render() {
    // @ts-ignore
    const { myId, person, iAmAdmin, iAmOwner, personIsAdmin } = this.props;

    const personIsMe = myId === person.id;

    const showLeaveCommunity = personIsMe;
    const showMakeAdmin = !personIsMe && iAmAdmin && !personIsAdmin;
    const showRemoveAsAdmin = !personIsMe && iAmAdmin && personIsAdmin;
    const showMakeOwner = !personIsMe && iAmOwner;
    const showRemoveFromCommunity = !personIsMe && iAmAdmin;

    const props = {
      actions: [
        ...(showLeaveCommunity ? this.canLeaveCommunity() : []),
        ...(showMakeAdmin
          ? this.createOption('makeAdmin', this.makeAdmin, true)
          : []),
        ...(showRemoveAsAdmin
          ? // prettier-ignore
            // @ts-ignore
            this.createOption('removeAdmin', this.removeAsAdmin)
          : []),
        ...(showMakeOwner
          ? this.createOption('makeOwner', this.makeOwner, true)
          : []),
        ...(showRemoveFromCommunity
          ? // prettier-ignore
            // @ts-ignore
            this.createOption('removeMember', this.removeFromCommunity)
          : []),
      ],
      buttonProps: { style: styles.container },
    };

    return <PopupMenu {...props} />;
  }
}

// @ts-ignore
MemberOptionsMenu.propTypes = {
  myId: PropTypes.string.isRequired,
  person: PropTypes.object.isRequired,
  organization: PropTypes.object.isRequired,
  personOrgPermission: PropTypes.object.isRequired,
  iAmAdmin: PropTypes.bool,
  iAmOwner: PropTypes.bool,
  personIsAdmin: PropTypes.bool,
};

export default connect()(MemberOptionsMenu);
