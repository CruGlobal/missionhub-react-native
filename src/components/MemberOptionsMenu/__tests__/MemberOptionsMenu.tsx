/* eslint max-lines: 0 */

import React from 'react';
import { Alert } from 'react-native';
import i18next from 'i18next';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { testSnapshotShallow, renderShallow } from '../../../../testUtils';
import {
  transferOrgOwnership,
  removeOrganizationMember,
} from '../../../actions/organizations';
import {
  makeAdmin,
  removeAsAdmin,
  archiveOrgPermission,
} from '../../../actions/person';
import { navigateBack } from '../../../actions/navigation';
import { PermissionEnum } from '../../../../__generated__/globalTypes';

import MemberOptionsMenu, {
  API_TRY_IT_NOW_ADMIN_OWNER_ERROR_MESSAGE,
} from '..';

jest.mock('../../../actions/organizations');
jest.mock('../../../actions/person');
jest.mock('../../../actions/navigation');

const myId = '1';
const otherId = '2';
const organization = { name: "Roge's org", id: '08747283423' };
const personOrgPermission = { id: '25234234', permission: PermissionEnum.user };

const person = { id: '1', full_name: 'Test Person' };
const mockStore = configureStore([thunk]);

// @ts-ignore
let props;
// @ts-ignore
let store;
const onActionTaken = jest.fn();
const defaultProps = { t: i18next.t, dispatch: () => {}, myId, onActionTaken };

const test = () => {
  // @ts-ignore
  testSnapshotShallow(<MemberOptionsMenu {...props} />);
};

describe('MemberOptionsMenu', () => {
  describe('for me, as owner', () => {
    beforeEach(
      () =>
        (props = {
          ...defaultProps,
          person: {
            ...person,
            id: myId,
          },
          iAmAdmin: false,
          iAmOwner: true,
          personIsAdmin: false,
          organization,
        }),
    );

    it('renders correctly', () => test());

    it('shows an alert message if I attempt to leave', () => {
      Alert.alert = jest.fn();
      // @ts-ignore
      const screen = renderShallow(<MemberOptionsMenu {...props} />);

      // @ts-ignore
      screen.props().actions[0].onPress();

      expect(Alert.alert).toHaveBeenCalledWith(
        i18next.t('groupMemberOptions:ownerLeaveCommunityErrorMessage', {
          orgName: organization.name,
        }),
        null,
        { text: i18next.t('ok') },
      );
    });
  });

  it('renders for admin looking at member', () => {
    props = {
      ...defaultProps,
      person: {
        ...person,
        id: otherId,
      },
      iAmAdmin: true,
      iAmOwner: false,
      personIsAdmin: false,
      organization,
    };
    test();
  });

  it('renders for admin looking at admin', () => {
    props = {
      ...defaultProps,
      person: {
        ...person,
        id: otherId,
      },
      iAmAdmin: true,
      iAmOwner: false,
      personIsAdmin: true,
      organization,
    };
    test();
  });

  describe(' looking at member, when I am owner', () => {
    beforeEach(
      () =>
        (props = {
          ...defaultProps,
          person: {
            ...person,
            id: otherId,
          },
          iAmAdmin: true,
          iAmOwner: true,
          personIsAdmin: false,
          organization,
        }),
    );

    it('renders correctly', () => test());

    it('transfers ownership', () => {
      // @ts-ignore
      transferOrgOwnership.mockReturnValue({ type: 'transferred ownership' });
      // @ts-ignore
      const screen = renderShallow(<MemberOptionsMenu {...props} />);

      // @ts-ignore
      screen.instance().makeOwner();

      expect(transferOrgOwnership).toHaveBeenCalledWith(
        organization.id,
        otherId,
      );
    });

    it('shows error message for Try It Now users', async () => {
      // @ts-ignore
      transferOrgOwnership.mockReturnValue(() =>
        Promise.reject({
          apiError: {
            errors: [
              {
                detail: 'blah ' + API_TRY_IT_NOW_ADMIN_OWNER_ERROR_MESSAGE,
              },
            ],
          },
        }),
      );

      // @ts-ignore
      await renderShallow(<MemberOptionsMenu {...props} />)
        .instance()
        // @ts-ignore
        .makeOwner();

      expect(Alert.alert).toHaveBeenCalledWith(
        i18next.t('groupMemberOptions:tryItNowAdminOwnerErrorMessage'),
      );
    });

    it('throws unexpected errors', async () => {
      expect.assertions(1);
      const error = {
        apiError: {
          errors: [
            {
              detail: 'SCOTTTTYYYYYYYYYYYYYYYYYYYYYY',
            },
          ],
        },
      };
      // @ts-ignore
      transferOrgOwnership.mockReturnValue(() => Promise.reject(error));

      try {
        // @ts-ignore
        await renderShallow(<MemberOptionsMenu {...props} />)
          .instance()
          // @ts-ignore
          .makeOwner();
      } catch (e) {
        expect(e).toEqual(error);
      }
    });
  });

  it('renders for owner looking at admin', () => {
    props = {
      ...defaultProps,
      person: {
        ...person,
        id: otherId,
      },
      iAmAdmin: true,
      iAmOwner: true,
      personIsAdmin: true,
      organization,
    };
    test();
  });
});

describe('confirm screen', () => {
  Alert.alert = jest.fn();

  // @ts-ignore
  let screen;

  describe('Make Admin', () => {
    const makeAdminResponse = { type: 'make admin' };

    beforeEach(() => {
      props = {
        ...defaultProps,
        person: {
          ...person,
          id: otherId,
        },
        iAmAdmin: true,
        iAmOwner: false,
        personIsAdmin: false,
        organization,
        personOrgPermission,
      };

      store = mockStore();
      screen = renderShallow(<MemberOptionsMenu {...props} />, store);
    });

    it('displays confirm screen', () => {
      // @ts-ignore
      makeAdmin.mockReturnValue(makeAdminResponse);

      // @ts-ignore
      screen.props().actions[0].onPress();

      expect(Alert.alert).toHaveBeenCalledWith(
        i18next.t('groupMemberOptions:makeAdmin:modalTitle', {
          personName: person.full_name,
          communityName: organization.name,
        }),
        i18next.t('groupMemberOptions:makeAdmin:modalDescription'),
        [
          {
            text: i18next.t('cancel'),
            style: 'cancel',
          },
          {
            text: i18next.t('groupMemberOptions:makeAdmin:confirmButtonText'),
            onPress: expect.any(Function),
          },
        ],
      );
    });

    it('calls makeAdmin action', async () => {
      // @ts-ignore
      makeAdmin.mockReturnValue(makeAdminResponse);

      // @ts-ignore
      await screen.instance().makeAdmin();

      // @ts-ignore
      expect(store.getActions()).toEqual([makeAdminResponse]);
      expect(makeAdmin).toHaveBeenCalledWith(otherId, personOrgPermission.id);
    });

    it('shows error message for Try It Now users', async () => {
      // @ts-ignore
      makeAdmin.mockReturnValue(() =>
        Promise.reject({
          apiError: {
            errors: [
              {
                detail: {
                  permission_id: [API_TRY_IT_NOW_ADMIN_OWNER_ERROR_MESSAGE],
                },
              },
            ],
          },
        }),
      );

      // @ts-ignore
      await screen.instance().makeAdmin();

      expect(Alert.alert).toHaveBeenCalledWith(
        i18next.t('groupMemberOptions:tryItNowAdminOwnerErrorMessage'),
      );
    });

    it('throws unexpected errors', async () => {
      expect.assertions(1);
      const error = {
        apiError: {
          errors: [
            {
              detail: 'SCOTTTTYYYYYYYYYYYYYYYYYYYYYY',
            },
          ],
        },
      };
      // @ts-ignore
      makeAdmin.mockReturnValue(() => Promise.reject(error));

      try {
        // @ts-ignore
        await screen.instance().makeAdmin();
      } catch (e) {
        expect(e).toEqual(error);
      }
    });
  });

  describe('Remove As Admin', () => {
    const removeAdminResponse = { type: 'remove admin' };
    // @ts-ignore
    removeAsAdmin.mockReturnValue(removeAdminResponse);

    beforeEach(() => {
      props = {
        ...defaultProps,
        person: {
          ...person,
          id: otherId,
        },
        iAmAdmin: true,
        iAmOwner: false,
        personIsAdmin: true,
        organization,
        personOrgPermission,
      };

      store = mockStore();
      screen = renderShallow(<MemberOptionsMenu {...props} />, store);
    });

    it('displays confirm screen', () => {
      // @ts-ignore
      screen.props().actions[0].onPress();

      expect(Alert.alert).toHaveBeenCalledWith(
        i18next.t('groupMemberOptions:removeAdmin:modalTitle', {
          personName: person.full_name,
          communityName: organization.name,
        }),
        null,
        [
          {
            text: i18next.t('cancel'),
            style: 'cancel',
          },
          {
            text: i18next.t('groupMemberOptions:removeAdmin:confirmButtonText'),
            onPress: expect.any(Function),
          },
        ],
      );
    });

    it('calls removeAdmin action', async () => {
      // @ts-ignore
      await screen.instance().removeAsAdmin();

      // @ts-ignore
      expect(store.getActions()).toEqual([removeAdminResponse]);
      expect(onActionTaken).toHaveBeenCalled();
      expect(removeAsAdmin).toHaveBeenCalledWith(
        otherId,
        personOrgPermission.id,
      );
    });
  });
});

describe('Leave Community', () => {
  const navigateBackResult = { type: 'navigated back' };

  beforeEach(() => {
    props = {
      ...defaultProps,
      person: {
        ...person,
        id: myId,
      },
      iAmAdmin: true,
      iAmOwner: false,
      personIsAdmin: false,
      organization,
      personOrgPermission,
    };

    store = mockStore();
  });

  it('sets flag to archive my permission on unmount', async () => {
    // @ts-ignore
    navigateBack.mockReturnValue(navigateBackResult);
    // @ts-ignore
    const screen = renderShallow(<MemberOptionsMenu {...props} />, store);

    // @ts-ignore
    await screen.instance().leaveCommunity();

    expect(onActionTaken).toHaveBeenCalled();
    // @ts-ignore
    expect(store.getActions()).toEqual([navigateBackResult]);
  });

  it('sends api request to archive my permission on unmount if leaveCommunityOnUnmount flag set', async () => {
    // @ts-ignore
    archiveOrgPermission.mockReturnValue(() => Promise.resolve());
    // @ts-ignore
    const screen = renderShallow(<MemberOptionsMenu {...props} />, store);

    // @ts-ignore
    screen.instance().leaveCommunityOnUnmount = true;
    // @ts-ignore
    await screen.instance().componentWillUnmount();

    expect(archiveOrgPermission).toHaveBeenCalledWith(
      myId,
      personOrgPermission.id,
    );
  });

  it('does nothing on unmount if leaveCommunityOnUnmount flag unset', async () => {
    // @ts-ignore
    const screen = renderShallow(<MemberOptionsMenu {...props} />, store);

    // @ts-ignore
    await screen.instance().componentWillUnmount();

    // @ts-ignore
    expect(store.getActions()).toEqual([]);
    expect(archiveOrgPermission).not.toHaveBeenCalled();
  });
});

describe('Remove from Community', () => {
  const archiveOrgPermissionResult = { type: 'archived permission' };
  const removePersonResult = { type: 'removed person' };

  beforeEach(() => {
    props = {
      ...defaultProps,
      person: {
        ...person,
        id: otherId,
      },
      iAmAdmin: true,
      iAmOwner: false,
      personIsAdmin: false,
      organization,
      personOrgPermission,
    };

    store = mockStore();
  });

  it("sends api request to archive person's permission", async () => {
    // @ts-ignore
    archiveOrgPermission.mockReturnValue(archiveOrgPermissionResult);
    // @ts-ignore
    removeOrganizationMember.mockReturnValue(removePersonResult);
    // @ts-ignore
    const screen = renderShallow(<MemberOptionsMenu {...props} />, store);

    // @ts-ignore
    await screen.instance().removeFromCommunity();

    // @ts-ignore
    expect(store.getActions()).toEqual([
      archiveOrgPermissionResult,
      removePersonResult,
    ]);
    expect(archiveOrgPermission).toHaveBeenCalledWith(
      otherId,
      personOrgPermission.id,
    );
    expect(removeOrganizationMember).toHaveBeenCalledWith(
      otherId,
      organization.id,
    );
  });
});
