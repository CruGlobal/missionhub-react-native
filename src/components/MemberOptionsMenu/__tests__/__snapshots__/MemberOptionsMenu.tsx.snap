// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberOptionsMenu for me, as owner renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 32,
    }
  }
  testID="IconButton"
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 18,
          },
          Array [
            Object {
              "margin": 10,
            },
            Array [
              Object {
                "color": "#B4B6BA",
                "margin": 0,
                "paddingHorizontal": 10,
              },
              undefined,
            ],
          ],
        ],
        Object {
          "fontFamily": "icomoon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`MemberOptionsMenu looking at member, when I am owner renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 32,
    }
  }
  testID="IconButton"
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 18,
          },
          Array [
            Object {
              "margin": 10,
            },
            Array [
              Object {
                "color": "#B4B6BA",
                "margin": 0,
                "paddingHorizontal": 10,
              },
              undefined,
            ],
          ],
        ],
        Object {
          "fontFamily": "icomoon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`MemberOptionsMenu renders for admin looking at admin 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 32,
    }
  }
  testID="IconButton"
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 18,
          },
          Array [
            Object {
              "margin": 10,
            },
            Array [
              Object {
                "color": "#B4B6BA",
                "margin": 0,
                "paddingHorizontal": 10,
              },
              undefined,
            ],
          ],
        ],
        Object {
          "fontFamily": "icomoon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`MemberOptionsMenu renders for admin looking at member 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 32,
    }
  }
  testID="IconButton"
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 18,
          },
          Array [
            Object {
              "margin": 10,
            },
            Array [
              Object {
                "color": "#B4B6BA",
                "margin": 0,
                "paddingHorizontal": 10,
              },
              undefined,
            ],
          ],
        ],
        Object {
          "fontFamily": "icomoon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`MemberOptionsMenu renders for owner looking at admin 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "height": 32,
      "justifyContent": "center",
      "opacity": 1,
      "width": 32,
    }
  }
  testID="IconButton"
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 18,
          },
          Array [
            Object {
              "margin": 10,
            },
            Array [
              Object {
                "color": "#B4B6BA",
                "margin": 0,
                "paddingHorizontal": 10,
              },
              undefined,
            ],
          ],
        ],
        Object {
          "fontFamily": "icomoon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;
